File: services/kfc-agent-backend/test/recommendations/automatic-recommendation-contract.test.ts
import { describe, expect, it } from 'vitest';
import { readFile } from 'node:fs/promises';

const contractModulePath =
  '../../src/recommendations/contracts/automatic-recommendation.js';

const commonRequest = {
  schemaVersion: 'kfc-automatic-recommendation-v1',
  requestId: 'request-contract-001',
  storeId: 'KFCVN0002',
  fulfilmentMode: 'pickup',
  locale: 'vi-VN',
  orderingJourneyRef: 'journey-contract-001',
  opportunityRef: 'opportunity-contract-001',
  cart: {
    cartId: 'cart-contract-001',
    revision: 'cart-revision-contract-001',
    subtotal: { amount: 0, currency: 'VND' },
    lines: [],
  },
} as const;

describe('automatic recommendation wire contract', () => {
  it('exposes four exact decision operations without a generic decide route', async () => {
    const contract = await import(contractModulePath);

    expect(contract.automaticRecommendationOperations).toEqual({
      local_favorite: '/v1/recommendations/local-favorites',
      for_you: '/v1/recommendations/for-you',
      modifier_upsell: '/v1/recommendations/modifier-upsells',
      smart_cross_sell: '/v1/recommendations/smart-cross-sells',
    });
  });

  it('binds canonical request and event identity digests to type and operation path', async () => {
    const { automaticRecommendationIdentityDigest } = await import(
      contractModulePath
    );
    const request = { cart: { revision: 'cart-1' }, storeId: 'KFCVN0002' };
    const reordered = { storeId: 'KFCVN0002', cart: { revision: 'cart-1' } };
    const input = {
      operationPath: '/v1/recommendations/local-favorites',
      identityType: 'local_favorite',
      payload: request,
    };
    expect(automaticRecommendationIdentityDigest(input)).toBe(
      automaticRecommendationIdentityDigest({ ...input, payload: reordered }),
    );
    expect(automaticRecommendationIdentityDigest(input)).not.toBe(
      automaticRecommendationIdentityDigest({
        ...input,
        identityType: 'smart_cross_sell',
      }),
    );
    expect(automaticRecommendationIdentityDigest(input)).not.toBe(
      automaticRecommendationIdentityDigest({
        ...input,
        operationPath: '/v1/recommendations/local-favorites/events',
      }),
    );
  });

  it('matches the published cross-runtime identity digest vector', async () => {
    const { automaticRecommendationIdentityDigest } = await import(
      contractModulePath
    );
    const manifest = await import(
      '../../../../contracts/automatic-recommendations/v1/contract-manifest.json',
      { with: { type: 'json' } }
    );
    const vector = manifest.default.identityDigestVector;

    expect(automaticRecommendationIdentityDigest(vector)).toBe(vector.sha256);
  });

  it('enforces each trusted type-specific request prerequisite', async () => {
    const { parseAutomaticRecommendationRequest } = await import(
      contractModulePath
    );

    expect(
      parseAutomaticRecommendationRequest('local_favorite', commonRequest),
    ).toMatchObject({
      orderingJourneyRef: 'journey-contract-001',
      opportunityRef: 'opportunity-contract-001',
    });
    expect(() =>
      parseAutomaticRecommendationRequest('for_you', commonRequest),
    ).toThrow();
    expect(() =>
      parseAutomaticRecommendationRequest('modifier_upsell', commonRequest),
    ).toThrow();
    expect(() =>
      parseAutomaticRecommendationRequest('smart_cross_sell', commonRequest),
    ).toThrow();

    expect(
      parseAutomaticRecommendationRequest('for_you', {
        ...commonRequest,
        verifiedCustomerRef: 'verified-customer-contract-001',
      }),
    ).toMatchObject({ verifiedCustomerRef: 'verified-customer-contract-001' });
    expect(
      parseAutomaticRecommendationRequest('modifier_upsell', {
        ...commonRequest,
        parentCartLineId: 'cart-line-contract-001',
      }),
    ).toMatchObject({ parentCartLineId: 'cart-line-contract-001' });
    expect(
      parseAutomaticRecommendationRequest('smart_cross_sell', {
        ...commonRequest,
        cart: {
          ...commonRequest.cart,
          subtotal: { amount: 89000, currency: 'VND' },
          lines: [
            {
              lineId: 'cart-line-contract-001',
              sellableItemId: '20732',
              quantity: 1,
              unitPrice: { amount: 89000, currency: 'VND' },
              modifiers: [],
            },
          ],
        },
      }),
    ).toMatchObject({
      cart: { lines: [{ lineId: 'cart-line-contract-001' }] },
    });
  });

  it('rejects client-authored scoring and authority fields', async () => {
    const { parseAutomaticRecommendationRequest } = await import(
      contractModulePath
    );

    for (const forbiddenField of [
      { candidates: [] },
      { features: {} },
      { modelRevision: 'client-model' },
      { threshold: 0 },
      { decisionTime: '2026-08-04T00:00:00Z' },
      { commerceSnapshotBindings: {} },
    ]) {
      expect(() =>
        parseAutomaticRecommendationRequest('local_favorite', {
          ...commonRequest,
          ...forbiddenField,
        }),
      ).toThrow();
    }
  });

  it('represents a governed pause without substituting fallback output', async () => {
    const { parseAutomaticRecommendationResponse } = await import(
      contractModulePath
    );

    const paused = parseAutomaticRecommendationResponse({
      schemaVersion: 'kfc-automatic-recommendation-v1',
      requestId: commonRequest.requestId,
      recommendationId: 'recommendation-contract-001',
      recommendationType: 'local_favorite',
      status: 'paused',
      emptyReason: 'recommendation_serving_paused',
      cartRevision: commonRequest.cart.revision,
      catalogRevision: 'catalog-revision-contract-001',
      expiresAt: '2026-08-04T13:30:00Z',
      model: null,
      proposals: [],
      counts: { potential: 0, eligible: 0, scored: 0, displayed: 0 },
    });

    expect(paused).toMatchObject({
      status: 'paused',
      model: null,
      proposals: [],
    });
    expect(() =>
      parseAutomaticRecommendationResponse({
        ...paused,
        fallbackSource: 'popularity',
      }),
    ).toThrow();
  });

  it('accepts rendered impressions and requires actions for selection outcomes', async () => {
    const {
      parseAutomaticRecommendationImpression,
      parseAutomaticRecommendationOutcome,
    } = await import(contractModulePath);

    const impression = {
      schemaVersion: 'kfc-automatic-recommendation-event-v1',
      eventId: 'event-impression-contract-001',
      channel: 'kiosk',
      occurredAt: '2026-08-04T13:00:00Z',
      orderingJourneyRef: 'journey-contract-001',
      opportunityRef: 'opportunity-contract-001',
      cartRevision: commonRequest.cart.revision,
      renderedActions: [{ actionId: 'product:20732', renderedPosition: 1 }],
    };
    expect(parseAutomaticRecommendationImpression(impression)).toMatchObject({
      renderedActions: [{ renderedPosition: 1 }],
    });

    expect(() =>
      parseAutomaticRecommendationOutcome({
        schemaVersion: 'kfc-automatic-recommendation-event-v1',
        eventId: 'event-outcome-contract-001',
        channel: 'kiosk',
        eventType: 'selected',
        occurredAt: '2026-08-04T13:01:00Z',
        orderingJourneyRef: 'journey-contract-001',
        opportunityRef: 'opportunity-contract-001',
        actionId: null,
        renderedPosition: 1,
        cartRevision: commonRequest.cart.revision,
      }),
    ).toThrow();

    expect(() =>
      parseAutomaticRecommendationOutcome({
        ...impression,
        eventId: 'event-outcome-contract-002',
        eventType: 'slate_dismissed',
        actionId: 'product:20732',
        renderedPosition: 1,
      }),
    ).toThrow();
  });

  it('accepts only eligible rows and reconciled bounded scorer output', async () => {
    const {
      parseAutomaticScorerRequest,
      parseAutomaticScorerResponse,
      reconcileAutomaticScorerResponse,
    } = await import(contractModulePath);
    const model = {
      bundleId: 'qualified-bundle-contract-001',
      bundleDigest:
        'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
      modelRevision: 'local-favorite-model-contract-001',
      calibratorRevision: 'local-favorite-calibrator-contract-001',
      featureSchemaDigest:
        'bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb',
      thresholdRevision: 'local-favorite-threshold-contract-001',
      composerContractDigest:
        'cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc',
      qualificationRunId: 'qualification-run-contract-001',
      qualificationEvidenceDigest:
        'dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd',
    };

    const scorerRequest = {
      schemaVersion: 'kfc-automatic-scorer-v1',
      requestId: commonRequest.requestId,
      recommendationType: 'local_favorite',
      model,
      candidates: [
        {
          candidateId: 'product:20732',
          eligibility: 'eligible',
          priceImpactVnd: 89000,
          features: { priceImpactVnd: 89000, daypart: 'lunch' },
        },
      ],
    };
    expect(parseAutomaticScorerRequest(scorerRequest)).toMatchObject({
      candidates: [{ eligibility: 'eligible' }],
    });

    const scorerResponse = {
      schemaVersion: 'kfc-automatic-scorer-v1',
      requestId: commonRequest.requestId,
      model,
      scores: [
        {
          candidateId: 'product:20732',
          selectionProbability: 0.4,
          jointProbability: 0.3,
          explanationValues: { localDemand: 0.25, priceImpact: -0.1 },
        },
      ],
    };
    expect(parseAutomaticScorerResponse(scorerResponse)).toMatchObject({
      scores: [{ jointProbability: 0.3 }],
    });
    expect(
      reconcileAutomaticScorerResponse(scorerRequest, scorerResponse),
    ).toMatchObject({
      scores: [{ candidateId: 'product:20732' }],
    });
    const scorerVectorRequest = JSON.parse(
      await readFile(
        new URL(
          '../../../../contracts/automatic-recommendations/v1/examples/scorer-request.json',
          import.meta.url,
        ),
        'utf8',
      ),
    );
    const reorderedModelResponse = JSON.parse(
      await readFile(
        new URL(
          '../../../../contracts/automatic-recommendations/v1/examples/scorer-reordered-model-response.json',
          import.meta.url,
        ),
        'utf8',
      ),
    );
    expect(
      reconcileAutomaticScorerResponse(
        scorerVectorRequest,
        reorderedModelResponse,
      ),
    ).toMatchObject({ scores: [{ candidateId: 'product:20732' }] });
    for (const invalidResponse of [
      { ...scorerResponse, requestId: 'different-request' },
      {
        ...scorerResponse,
        model: { ...model, modelRevision: 'different-model' },
      },
      { ...scorerResponse, scores: [] },
      {
        ...scorerResponse,
        scores: [...scorerResponse.scores, scorerResponse.scores[0]],
      },
      {
        ...scorerResponse,
        scores: [{ ...scorerResponse.scores[0], candidateId: 'extra-product' }],
      },
    ]) {
      expect(() =>
        reconcileAutomaticScorerResponse(scorerRequest, invalidResponse),
      ).toThrow();
    }

    expect(() =>
      parseAutomaticScorerResponse({
        schemaVersion: 'kfc-automatic-scorer-v1',
        requestId: commonRequest.requestId,
        model,
        scores: [
          {
            candidateId: 'product:20732',
            selectionProbability: 0.2,
            jointProbability: 0.3,
            explanationValues: {},
          },
        ],
      }),
    ).toThrow();
  });

  it('binds modifier actions to the requested parent cart line', async () => {
    const { validateAutomaticRecommendationBinding } = await import(
      contractModulePath
    );
    const root = new URL(
      '../../../../contracts/automatic-recommendations/v1/examples/',
      import.meta.url,
    );
    const request = JSON.parse(
      await readFile(new URL('modifier-upsell-request.json', root), 'utf8'),
    );
    const mismatchedResponse = JSON.parse(
      await readFile(
        new URL('adversarial/modifier-parent-mismatch-response.json', root),
        'utf8',
      ),
    );

    expect(() =>
      validateAutomaticRecommendationBinding(
        'modifier_upsell',
        request,
        mismatchedResponse,
      ),
    ).toThrow();
  });

  it('requires trusted journey and opportunity references without client ranking authority', async () => {
    const { parseAutomaticRecommendationRequest } = await import(
      contractModulePath
    );

    expect(() =>
      parseAutomaticRecommendationRequest('local_favorite', {
        ...commonRequest,
        orderingJourneyRef: undefined,
      }),
    ).toThrow();
  });
});

File: apps/kfc_live_monitor_flutter/test/features/automatic_recommendations/automatic_recommendation_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kfc_live_monitor/features/automatic_recommendations/domain/automatic_recommendation_contract.dart';

void main() {
  test('Dart client exposes the four exact decision operations', () {
    expect(
      automaticRecommendationSchemaVersion,
      'kfc-automatic-recommendation-v1',
    );
    expect(automaticRecommendationOperationPaths, {
      AutomaticRecommendationType.localFavorite:
          '/v1/recommendations/local-favorites',
      AutomaticRecommendationType.forYou: '/v1/recommendations/for-you',
      AutomaticRecommendationType.modifierUpsell:
          '/v1/recommendations/modifier-upsells',
      AutomaticRecommendationType.smartCrossSell:
          '/v1/recommendations/smart-cross-sells',
    });
  });

  test('Dart representation matches the canonical contract digest', () {
    final contractRoot = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/',
    );
    final manifest =
        jsonDecode(
              File.fromUri(
                contractRoot.resolve('contract-manifest.json'),
              ).readAsStringSync(),
            )
            as Map<String, Object?>;
    final authorityFiles = (manifest['authorityFiles']! as List<Object?>)
        .cast<String>();
    final bytes = <int>[];

    for (final relativePath in authorityFiles) {
      bytes
        ..addAll(utf8.encode(relativePath))
        ..add(0)
        ..addAll(
          File.fromUri(contractRoot.resolve(relativePath)).readAsBytesSync(),
        )
        ..add(0);
    }

    expect(
      automaticRecommendationContractDigest,
      sha256.convert(bytes).toString(),
    );
    expect(automaticRecommendationContractDigest, manifest['canonicalDigest']);
  });

  test('Dart representations parse every canonical fixture', () {
    final contractRoot = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/',
    );
    final manifest =
        jsonDecode(
              File.fromUri(
                contractRoot.resolve('contract-manifest.json'),
              ).readAsStringSync(),
            )
            as Map<String, Object?>;
    final parsers = <String, AutomaticRecommendationPayload Function(Object?)>{
      'local_favorite_request': (value) =>
          AutomaticRecommendationRequestPayload.parse(
            AutomaticRecommendationType.localFavorite,
            value,
          ),
      'for_you_request': (value) => AutomaticRecommendationRequestPayload.parse(
        AutomaticRecommendationType.forYou,
        value,
      ),
      'modifier_upsell_request': (value) =>
          AutomaticRecommendationRequestPayload.parse(
            AutomaticRecommendationType.modifierUpsell,
            value,
          ),
      'smart_cross_sell_request': (value) =>
          AutomaticRecommendationRequestPayload.parse(
            AutomaticRecommendationType.smartCrossSell,
            value,
          ),
      'recommendation_response': AutomaticRecommendationResponsePayload.parse,
      'impression_request': AutomaticRecommendationImpressionPayload.parse,
      'outcome_request': AutomaticRecommendationOutcomePayload.parse,
      'problem_details': AutomaticRecommendationProblemPayload.parse,
      'inspection_response': AutomaticRecommendationInspectionPayload.parse,
      'scorer_request': AutomaticScorerRequestPayload.parse,
      'scorer_response': AutomaticScorerResponsePayload.parse,
    };

    for (final example in (manifest['examples']! as List<Object?>)) {
      final descriptor = example! as Map<String, Object?>;
      final value = jsonDecode(
        File.fromUri(
          contractRoot.resolve(descriptor['file']! as String),
        ).readAsStringSync(),
      );
      expect(parsers[descriptor['kind']]!(value).toJson(), value);
    }
  });

  test('Dart representations reject every negative fixture', () {
    final contractRoot = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/',
    );
    final negativeExamples =
        <String, AutomaticRecommendationPayload Function(Object?)>{
          'examples/negative/request-missing-journey-reference.json': (value) =>
              AutomaticRecommendationRequestPayload.parse(
                AutomaticRecommendationType.localFavorite,
                value,
              ),
          'examples/negative/outcome-generic-payload.json':
              AutomaticRecommendationOutcomePayload.parse,
          'examples/negative/scorer-missing-provenance.json':
              AutomaticScorerRequestPayload.parse,
          'examples/negative/problem-status-code-mismatch.json':
              AutomaticRecommendationProblemPayload.parse,
          'examples/adversarial/scorer-nested-feature.json':
              AutomaticScorerRequestPayload.parse,
          'examples/adversarial/recommended-invented-reason.json':
              AutomaticRecommendationResponsePayload.parse,
          'examples/adversarial/recommended-without-model.json':
              AutomaticRecommendationResponsePayload.parse,
          'examples/adversarial/problem-503-not-retryable.json':
              AutomaticRecommendationProblemPayload.parse,
          'examples/adversarial/modifier-with-product-action.json':
              AutomaticRecommendationResponsePayload.parse,
          'examples/adversarial/impression-empty.json':
              AutomaticRecommendationImpressionPayload.parse,
          'examples/adversarial/recommended-nonmonotonic-counts.json':
              AutomaticRecommendationResponsePayload.parse,
          'examples/adversarial/modifier-four-proposals.json':
              AutomaticRecommendationResponsePayload.parse,
        };

    for (final entry in negativeExamples.entries) {
      final value = jsonDecode(
        File.fromUri(contractRoot.resolve(entry.key)).readAsStringSync(),
      );
      expect(
        () => entry.value(value),
        throwsA(isA<AutomaticRecommendationContractException>()),
      );
    }
  });

  test('Dart scorer reconciliation requires exact pairing', () {
    final root = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/examples/',
    );
    final request =
        jsonDecode(
              File.fromUri(
                root.resolve('scorer-request.json'),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    final response =
        jsonDecode(
              File.fromUri(
                root.resolve('scorer-response.json'),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    expect(
      reconcileAutomaticScorerResponse(request, response).toJson(),
      response,
    );
    final reorderedModelResponse =
        jsonDecode(
              File.fromUri(
                root.resolve('scorer-reordered-model-response.json'),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    expect(
      reconcileAutomaticScorerResponse(
        request,
        reorderedModelResponse,
      ).toJson(),
      reorderedModelResponse,
    );
    for (final invalid in [
      {...response, 'requestId': 'mismatch'},
      {...response, 'scores': []},
      {
        ...response,
        'scores': [response['scores']![0], response['scores']![0]],
      },
      {
        ...response,
        'scores': [
          {...response['scores']![0], 'candidateId': 'extra'},
        ],
      },
    ]) {
      expect(
        () => reconcileAutomaticScorerResponse(request, invalid),
        throwsA(isA<AutomaticRecommendationContractException>()),
      );
    }
  });

  test('Dart scorer feature vector is fixed and type-applicable', () {
    final root = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/examples/',
    );
    final request =
        jsonDecode(
              File.fromUri(
                root.resolve('scorer-request.json'),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    for (final mutate in <void Function(Map<String, dynamic>)>[
      (features) => features.remove('storeId'),
      (features) => features['selectedAfterDisplay'] = true,
      (features) => features['modifierGroupPath'] = 'meal/side',
    ]) {
      final invalid = jsonDecode(jsonEncode(request)) as Map<String, dynamic>;
      final candidates = invalid['candidates'] as List<dynamic>;
      final candidate = candidates.single as Map<String, dynamic>;
      mutate(candidate['features'] as Map<String, dynamic>);
      expect(
        () => AutomaticScorerRequestPayload.parse(invalid),
        throwsA(isA<AutomaticRecommendationContractException>()),
      );
    }
  });

  test('Dart identity digest binds type and path', () {
    final request = {
      'cart': {'revision': 'cart-1'},
      'storeId': 'KFCVN0002',
    };
    final digest = automaticRecommendationIdentityDigest(
      operationPath: '/v1/recommendations/local-favorites',
      identityType: 'local_favorite',
      payload: request,
    );
    expect(
      digest,
      automaticRecommendationIdentityDigest(
        operationPath: '/v1/recommendations/local-favorites',
        identityType: 'local_favorite',
        payload: {
          'storeId': 'KFCVN0002',
          'cart': {'revision': 'cart-1'},
        },
      ),
    );
    expect(
      digest,
      isNot(
        automaticRecommendationIdentityDigest(
          operationPath: '/v1/recommendations/local-favorites',
          identityType: 'smart_cross_sell',
          payload: request,
        ),
      ),
    );
  });

  test('Dart identity digest matches the published cross-runtime vector', () {
    final contractRoot = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/',
    );
    final manifest =
        jsonDecode(
              File.fromUri(
                contractRoot.resolve('contract-manifest.json'),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    final vector = manifest['identityDigestVector'] as Map<String, dynamic>;

    expect(
      automaticRecommendationIdentityDigest(
        operationPath: vector['operationPath'] as String,
        identityType: vector['identityType'] as String,
        payload: vector['payload'],
      ),
      vector['sha256'],
    );
  });

  test('Dart modifier binding requires the requested parent cart line', () {
    final root = Directory.current.parent.parent.uri.resolve(
      'contracts/automatic-recommendations/v1/examples/',
    );
    final request =
        jsonDecode(
              File.fromUri(
                root.resolve('modifier-upsell-request.json'),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    final mismatch =
        jsonDecode(
              File.fromUri(
                root.resolve(
                  'adversarial/modifier-parent-mismatch-response.json',
                ),
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;

    expect(
      () => validateAutomaticRecommendationBinding(
        AutomaticRecommendationType.modifierUpsell,
        request,
        mismatch,
      ),
      throwsA(isA<AutomaticRecommendationContractException>()),
    );
  });
}
